### PR TITLE
Add sourcetype to Program node in ES6 (closes #46)

### LINKED ISSUE #46: Add flag on Program node to indicate module code?
It would be nice to be able to tell if a given Program represents module code or not. The Shift AST actually has two different nodes `Script` and `Module` to represent the source type. It's probably too big of a change to go for a separate node in ESTree, however, what about a property on `Program`? Maybe `sourceType` as "script" or "module"?

### PR DESCRIPTION
**Pull request by nzakas**
Adding `sourceType` to `Program` node.

### CODE REVIEW

**Comment by caridy** [comment]
:+1:

**Comment by getify** [comment]
+1

File: es6.md [hunk 1]
```diff
@@ -1,5 +1,15 @@
 This document specifies the extensions to the core ESTree AST types to support the ES6 grammar.
 
+# Programs
+
+```js
+extend interface Program {
+    sourceType: "script" | "module"
```
Comment (RReverser) [inline es6.md]: Grammar nazi mode: please put a semicolon here :)

Other than that, :+1:

Comment (nzakas) [inline es6.md]: Heh, my pleasure.

**Comment by nzakas** [comment]
Updated. Too bad there's not a linter. ;-)

**Comment by nzakas** [comment]
Ready for merge?

**Comment by RReverser** [comment]
Probably needs feedback from @ariya and @dherman.

**Comment by mikesherov** [comment]
:+1:

**Comment by michaelficarra** [comment]
:+1:

**Comment by jeffmo** [comment]
Big :+1:

**Comment by jeffmo** [comment]
In that past I'd also considered the thought of just having a `Program` (for backward compat) and a `Module` toplevel container. I suppose this works just as well, it just means the limitations on the contained statements in the body of `Program` has to be based on the property of the toplevel object (rather than its type).

**Comment by jeffmo** [comment]
FWIW it's likely that modules will continue to have a unique set of valid statements they can contain -- for example there's discussion about `await import` brewing (and toplevel `await` in general in a module) since modules are loaded asynchronously already.

**Comment by RReverser** [comment]
Well, yeah, but obviously it's not the only place where allowed set is different - for example, any `Function` node with `generator: true` can contain `YieldExpression`s, while any other can't.

This one doesn't seem to harm any processing / checks, so `sourceType` looks reasonable to me.

**Comment by jeffmo** [comment]
That's true. To be clear: I'm not suggesting you can necessarily do something with one over the other (nor am I suggesting that I'm in any way opposed to this PR) -- just throwing the differences out there in case it flags something for other consumers.

**Comment by dherman** [comment]
:+1: -- I had the same initial instinct as @jeffmo but I think this is an elegant way to pretend I meant to do it this way all along. ;-P

**Comment by mikesherov** [comment]
Thanks everyone!

**Pull request merged**